File: relevance/kwrt-packages/luci-app-xclient/luasrc/model/cbi/xclient/dns.lua
local m, s, o
local xclient = "xclient"
local uci = luci.model.uci.cursor()

m = Map(xclient)

s = m:section(TypedSection, "global", "DNS Servers")
s.anonymous = true
s.addremove   = false

o = s:option(DynamicList, "dns_servers", "Servers")
o.rmempty = false

o = s:option(Value, 'clientIp', 'Client Ip')
o.rmempty = true

o = s:option(ListValue, 'queryStrategy', 'Query Strategy')
o:value("UseIP", "UseIP")
o:value("UseIPv4", "UseIPv4")
o:value("UseIPv6", "UseIPv6")
o.rmempty = true


o = s:option(ListValue, 'disableCache', 'Disable Cache')
o:value("0", "False")
o:value("1", "True")
o.rmempty = true

o = s:option(ListValue, 'disableFallback', 'Disable Fallback')
o:value("0", "False")
o:value("1", "True")
o.rmempty = true

o = s:option(ListValue, 'disableFallbackIfMatch', 'Disable Fallback If Match')
o:value("0", "False")
o:value("1", "True")
o.rmempty = true
o.description = "Refer To Documentation"
..translatef("<a href=\"%s\" target=\"_blank\">" .. "%s".."</a>", "https://xtls.github.io/Xray-docs-next/config/dns.html", " Here")

y = Map("xclient")
y.pageaction = true

x = y:section(TypedSection, "dns", "DNS Rules")
x.anonymous = true
x.addremove = true
x.sortable = true

x.template = "cbi/tblsection"
x.extedit = luci.dispatcher.build_url("admin/services/xclient/add-dns/%s")
function x.create(...)
	local sid = TypedSection.create(...)
	if sid then
		luci.http.redirect(x.extedit % sid)
		return
	end
end


o = x:option(DummyValue, "address", "Address")
function o.cfgvalue(...)
	return Value.cfgvalue(...) or "—"
end

o = x:option(DummyValue, "port", "Port")
function o.cfgvalue(...)
	return Value.cfgvalue(...) or "—"
end


o = x:option(DummyValue, "domains", "Domains")
o.rawhtml=true
function o.cfgvalue(t,e)
local t,e=t.map:get(e,"domains"),""
if t then
for a,t in pairs(t)do
e=e..t.."<br>"
end
return e
else
return"&#8212;"
end
end

o = x:option(DummyValue, "expectIPs", "ExpectIPs")
o.rawhtml=true
function o.cfgvalue(t,e)
local t,e=t.map:get(e,"expectIPs"),""
if t then
for a,t in pairs(t)do
e=e..t.."<br>"
end
return e
else
return"&#8212;"
end
end
	
o = x:option(Flag, "skipFallback", "SkipFallback")
o.rmempty     = false
o.default     = o.skipFallback
o.cfgvalue    = function(...)
    return Flag.cfgvalue(...) or "0"
end


local apply = luci.http.formvalue("cbi.apply")
if apply then
  m.uci:commit("xclient")
  y.uci:commit("xclient")
  luci.sys.call("/etc/init.d/xclient boot >/dev/null 2>&1 &")
end

return m,y

File: luci-app-xjay/files/luci/dns.js
'use strict';
'require view';
'require form';
'require xjay';

return view.extend({

    render:function(){
        var m, s, o, ss;
        m = new form.Map('xjay', _('DNS'), _('DNS settings for xjay.'));

        s = m.section(form.TypedSection, 'dns');
        s.addremove = false;
        s.anonymous = true;

        o = s.option(form.ListValue, 'querystrategy', _('DNS Query Strategy'), _("To query only IPv4 or IPv6 or both."));
        o.value("UseIP");
        o.value("UseIPv4");
        o.value("UseIPv6");
        o.optional = true;

        o = s.option(form.Flag, 'disablecache', _('DNS Cache'), _('Enable DNS cache may speed up dns resolving speed.'));
        o.enabled = 'false';
        o.disabled = 'true';

        o = s.option(form.Flag, 'disablefallback', _('Disable Fallback Query'), _('Stop failed dns query to query using fallback servers again.'));
        o.enabled = 'true';
        o.disabled = 'false';

        o = s.option(form.Flag, 'disablefallbackifmatch', _('Disable Fallback If Match'), _('Disable fallback query when dns query match one domain in the DNS domain list.'));
        o.enabled = 'true';
        o.disabled = 'false';

        o = s.option(form.Value, "clientip", _("Client IP Address"));
        o.datatype = 'ipaddr';
        o.placeholder = '124.23.35.22';

        o = s.option(form.Value, "tag", _("Tag"), _('For routing rules to match the traffic sent by built-in DNS.'));
        o.placeholder = 'dns_inbound';

        o = s.option(form.DynamicList, "host", _("Hosts"), _("Set static domains. Queries of these domains will get the IP address mapped to them"));
        o.validate = xjay.validateServerSpec;
        o.placeholder = "/dns.google/8.8.8.8";
        o.rmempty = true;

        o = s.option(form.DynamicList, 'alt_dns', _('Alternative DNS Servers'), _("Using these DNS server to query the domains that doen't match any other dns server domains."));
        o.validate = xjay.validateDNS;
        o.placeholder = "8.8.8.8";
        o.rmempty = true;

        o = s.option(form.SectionValue, "dns_servers", form.GridSection, 'dns_server', _('DNS Servers'), _('DNS servers for rule based queries. Only domains that server domain list will using that DNS server to query. See <a href="https://xtls.github.io/config/dns.html#dns-%E6%9C%8D%E5%8A%A1%E5%99%A8">Built-in DNS server</a> for details.'));
        ss = o.subsection;
        ss.sortable = false;
        ss.anonymous = true;
        ss.addremove = true;
        ss.nodescriptions = true;

        o = ss.option(form.Value, "tag", _("Tag"), _("Alias name for this dns server."));
        o.validate = xjay.validateStingWhitespace;
        o.rmempty = false;

        o = ss.option(form.ListValue, 'querystrategy', _('DNS Query Strategy'), _("To query only IPv4 or IPv6 or both."));
        o.value("UseIP");
        o.value("UseIPv4");
        o.value("UseIPv6");
        o.optional = true;

        o = ss.option(form.Value, "server_address",_("Server IP Address"),  _("Valid server could be like 8.8.8.8 or 8.8.8.8:5353 or https://dns.google/dns-query or https+local://dns.google/dns-query or tcp://8.8.8.8:53 or tcp+local://8.8.8.8:53 or localhost."));
        o.validate = xjay.validateDNS;
        o.placeholder = "8.8.8.8";
        o.rmempty = false;

        o = ss.option(form.Value, "server_port", _("Server Port"), _("DNS server port to query domains."));
        o.datatype = "port"
        o.placeholder = '53'
        o.rmempty = false;

        o = ss.option(form.DynamicList, "server_domain", _("Domains"), _('Specify rules like <code>geosite:cn</code> or <code>domain:bilibili.com</code>. See <a href="https://xtls.github.io/config/dns.html#dnsobject">documentation</a> for details.'));
        o.validate = xjay.validateRoutingDomains;
        o.placeholder = 'geosite:google';
        o.datatype = "string";
        o.modalonly = true;

        o = ss.option(form.DynamicList, "server_expectedip", _("Expected IP Address"), _('Specify IP address so that only query result that matches on of the IP addresses will be returned.'));
        o.validate = xjay.validateRoutingIP;
        o.placeholder = 'geoip:cn';
        o.modalonly = true;

        o = ss.option(form.DynamicList, "server_clientip", _("Client IP Address"));
        o.datatype = 'ipaddr';
        o.placeholder = '124.23.35.22';
        o.modalonly = true;

        return m.render();
    }

});
